src/ray/gcs/gcs_server/gcs_autoscaler_state_manager.cc

```diff
@@ -276,7 +280,34 @@ void GcsAutoscalerStateManager::GetNodeStates(
 void GcsAutoscalerStateManager::HandleDrainNode(
     rpc::autoscaler::DrainNodeRequest request,
     rpc::autoscaler::DrainNodeReply *reply,
-    rpc::SendReplyCallback send_reply_callback) {}
+    rpc::SendReplyCallback send_reply_callback) {
+  const NodeID node_id = NodeID::FromBinary(request.node_id());
+  RAY_LOG(INFO) << "HandleDrainNode " << node_id.Hex()
+                << ", reason: " << request.reason_message();
+
+  auto node = gcs_node_manager_.GetAliveNode(node_id);
+  if (!node.has_value()) {
+    // The node is dead so treat it as drained.
+    reply->set_is_accepted(true);
+    send_reply_callback(ray::Status::OK(), nullptr, nullptr);
+    return;
+  }
+
+  rpc::Address raylet_address;
+  raylet_address.set_raylet_id(node.value()->node_id());
+  raylet_address.set_ip_address(node.value()->node_manager_address());
+  raylet_address.set_port(node.value()->node_manager_port());
+
+  const auto raylet_client = raylet_client_pool_->GetOrConnectByAddress(raylet_address);
+  raylet_client->DrainRaylet(
```

Comment:
would this ever fail?

Reply:
I think it can: for example, the raylet crashes.